src/machine.rs

```diff
@@ -0,0 +1,148 @@
+use std::collections::HashMap;
+use std::mem::align_of;
+
+use rustc::mir::{Mir, Local};
+
+use interp::TypedVal;
+
+const X64_ALIGN_SIZE: u8 = 8;
+const X32_ALIGN_SIZE: u8 = 4;
+
+#[derive(Debug)]
+struct Frame<'tcx> {
+    // MIR is a CFG of a single function in rust, so it makes sense
+    // for this to be in the frame.
+    mir: Mir<'tcx>,
+
+    // FIXME: Hacky... I can't find a way to actually access the private u32
+    // inside the `Local` tuple struct. `Local` does  implement the Hashable
+    // and PartialEQ trait however, so for  now it might just work to to use it
+    // as the key
+    locals: HashMap<Local, Vec<u8>>
+}
+
+impl<'tcx> Frame<'tcx> {
+    fn new(mir: Mir<'tcx>) -> Frame {
+        Frame {
+            locals: HashMap::new(),
+            mir:    mir
+        }
+    }
+
+    fn set_local(&mut self, local: Local, val: Vec<u8>) {
+        self.locals.insert(local, val);
+    }
+
+    fn get_local(&self, local: Local) -> &Vec<u8> {
+        self.locals.get(&local).unwrap()
+    }
+}
+
+// Represents a pointer into some kind of memory
+// TODO: Extend to work with static constructs
+enum Address {
+    Heap(usize),
+    Local(Local)
+}
+
+
+struct Memory {
+    // Memory is a simple vector of bytes whose size is parameterized at
+    // instantiation. The memory struct only cares about preserving alignment
+    // when handing out ptrs into it at runtime. Interpreting stored memory
+    // properly is the responsibility of those interfacing with it.
+    memory: Vec<u8>,
+
+    // Points to the last position in memory which is actually in use.
+    // Do not use directly to get a fresh ptr into memory.
+    // Instead use `next_aligned_ptr` method for a fresh correctly aligned ptr
+    // into memory for the target architecture.
+    next_free: usize,
+
+    target_alignment: u8,
+}
+
+impl Memory {
+    pub fn new(size: usize) -> Memory {
+        let alignment = align_of::<Vec>();
+        Memory {
+            memory: vec![0; size],
+            next_free: 0,
+            target_alignment: alignment
+        }
+    }
+
+    // Returns a pointer to next free block of memory of size n. Begins at index
+    // of the heap pointer.
+    //
+    // TODO: There is no free list so once a block of memory is allocated it is
+    // not yet possible to free it.
+    pub fn next_aligned_ptr(&mut self, size: usize) -> usize {
+        if (self.next_free + size) >= self.memory.len() {
+            panic!("Out of memory.")
+        }
+
+        let r = self.next_free % self.target_alignment;
+        if r > 0 {
+            self.next_free + r
+        }
+        else {
+            self.next_free
+        }
+    }
+
+    fn write(&mut self, ptr: usize, value: &[u8]) {
```

Comment:
I'm not sure `write` is the clearest name. Or, at least, put a brief doc above the function to make clear what its purpose is? [I don't have a suggestions for a better name, I admit, but I was a bit confused when I first looked at this function, as it wasn't clear from the function paramaters what was being written to what.]

Reply:
I've changed this to `store`. In my mind this might be slightly clearer, what do you think?

Reply:
Works for me.